gen/CDP/Gen/Program.hs

```diff
@@ -190,61 +186,100 @@ genCommand ctx domainName commandElt = T.unlines $
-    [ T.unlines enumDecls
-    , T.unwords ["data", paramsTypeName, "=", paramsTypeName, "{"]
-    , T.intercalate ",\n" fields
-    , "} " <> derivingGeneric
-    , genToJSONInstance paramsTypeName
-    , genFromJSONInstance paramsTypeName
-    ]
+genRecordType :: Context -> T.Text -> T.Text -> [D.Property] -> T.Text
+genRecordType ctx domainName recName params = T.unlines $
+    [ genTypeEnum ctx (tyNameHS "" (fieldNameHS recName $ D.propertyName p)) e
+    | p <- params, e <- maybeToList $ D.propertyEnum p
+    ] ++
+    [ "data " <> recName <> " = " <> recName] ++
+    (if null fields
+        then []
+        else ["  {"] ++ [T.intercalate ",\n" fields] ++ ["  }"]) ++
+    ["  deriving (Eq, Show)"]
   where
     fields = formatFieldDescription . map toField $ fieldSigDecls
-    toField ((fn,ftn,_),descM) = (T.unwords [fn, "::", ftn], descM)
-
-    enumDecls    = catMaybes . map (\((_,_,d),_) -> d) $ fieldSigDecls
-    fieldSigDecls = map (peltToFieldSigDecl &&& D.parametersEltDescription) paramElts
-
-    peltToFieldSigDecl pelt = case D.parametersEltEnum pelt of
-        Just enumValues -> (fn, ftn, ) . Just $ genTypeEnum ctx ftn enumValues
-        Nothing         -> (fn, , Nothing) $ genEltType ctx domainName (D.parametersEltOptional pelt == Just True)
-                (D.parametersEltType pelt)
-                (D.parametersEltRef pelt) 
-                (D.parametersEltItems pelt)
-      where
-        ftn = tyNameHS "" fn
-        fn = fieldNameHS paramsTypeName . D.parametersEltName $ pelt
+    toField ((fn,ftn),descM) = (T.unwords [fn, "::", ftn], descM)
+
+    fieldSigDecls = map (propertyHsSig ctx domainName recName &&& D.propertyDescription) params
+
+genRecordSmartConstructor :: Context -> T.Text -> T.Text -> [D.Property] -> T.Text
+genRecordSmartConstructor ctx domainName recName props = T.unlines $
+    [conName] ++
+    (do
+        (isFirstParam, (mbDesc, ty)) <- markFirst types
+        [ formatDescription 3 d | d <- maybeToList mbDesc ] ++
+            [ (if isFirstParam then "  :: " else "  -> ") <> ty ]) ++
+    [conName] ++
+    (map ("  " <>) args) ++
+    ["    = " <> recName] ++
+    (do
+        prop <- props
+        pure $ if D.propertyOptional prop
+            then "      Nothing"
+            else "      " <> argName prop)
+  where
+    conName = uncapitalizeFirst recName
+    requiredProps = filter (not . D.propertyOptional) props
+    types = (do
+        prop <- requiredProps
+        let (_, ty) = propertyHsSig ctx domainName recName prop
+        pure (D.propertyDescription prop, ty)) ++
+        [ (Nothing, recName) ]
+    args = do
+        prop <- requiredProps
+        guard . not $ D.propertyOptional prop
+        pure $ argName prop
+
+    argName = ("arg_" <>) . fieldNameHS recName . D.propertyName
+    markFirst = zip (True : repeat False)
+
+genRecordFromJson :: T.Text -> [D.Property] -> T.Text
+genRecordFromJson recName params = T.unlines $
```

Comment:
`genRecordFromJson` and `genRecordToJson` might benefit from the `space` function to clarify the indentation